presto-main/src/main/java/com/facebook/presto/execution/PlanFlattener.java

```diff
@@ -0,0 +1,269 @@
+        }
+    }
+
+    public static class FlattenedPlanFragment
+    {
+        private final String textPlan;
+        private final List<FlattenedNode> nodes;
+        private final PlanFragment fragment;
+
+        public static FlattenedPlanFragment fromPlanFragment(PlanFragment fragment, QueryExplainer explainer, Session session, FlatteningVisitor visitor)
+        {
+            ImmutableList.Builder<FlattenedNode> nodes = ImmutableList.builder();
+            fragment.getRoot().accept(visitor, nodes);
+            return new FlattenedPlanFragment(explainer.getPlan(fragment, session), fragment, nodes.build());
+        }
+
+        private FlattenedPlanFragment(String textPlan, PlanFragment fragment, List<FlattenedNode> nodes)
+        {
+            this.textPlan = requireNonNull(textPlan, "textPlan is null");
+            this.fragment = requireNonNull(fragment, "fragment is null");
+            this.nodes = ImmutableList.copyOf(requireNonNull(nodes, "nodes is null"));
+        }
+
+        @JsonProperty
+        public PlanFragmentId getId()
+        {
+            return fragment.getId();
+        }
+
+        @JsonProperty
+        public String getTextPlan()
+        {
+            return textPlan;
+        }
+
+        @JsonProperty
+        public PlanNode getTree()
+        {
+            return fragment.getRoot();
+        }
+
+        @JsonProperty
+        public List<FlattenedNode> getNodes()
+        {
+            return nodes;
+        }
+
+        @JsonProperty
+        public Map<Symbol, Type> getSymbols()
+        {
+            return fragment.getSymbols();
+        }
+
+        @JsonProperty
+        public PartitioningHandle getPartitioning()
+        {
+            return fragment.getPartitioning();
+        }
+
+        @JsonProperty
+        public List<PlanNodeId> getPartitionedSources()
+        {
+            return fragment.getPartitionedSources();
+        }
+
+        @JsonProperty
+        public List<Type> getTypes()
+        {
+            return fragment.getTypes();
+        }
+
+        @JsonProperty
+        public Set<PlanNode> getPartitionedSourceNodes()
+        {
+            return fragment.getPartitionedSourceNodes();
+        }
+
+        @JsonProperty
+        public List<RemoteSourceNode> getRemoteSourceNodes()
+        {
+            return fragment.getRemoteSourceNodes();
+        }
+
+        @JsonProperty
+        public PartitioningScheme getPartitioningScheme()
+        {
+            return fragment.getPartitioningScheme();
+        }
+    }
+
+    public static class FlattenedNode
+    {
+        private final String node;
+
+        public FlattenedNode(String node)
+        {
+            this.node = node;
+        }
+
+        @JsonValue
+        @JsonRawValue
+        public String getNode()
+        {
+            return node;
+        }
+    }
+
+    private static class FlatteningVisitor
+            extends SimplePlanVisitor<ImmutableList.Builder<FlattenedNode>>
```

Comment:
This is a bit unfortunate, since I'm trying to get rid of plan visitors in general (new planner rules don't need them, for instance)

Reply:
:(
How would one traverse a plan without a visitor?